### PR TITLE
Fix late-binding closure in `configure_autologging_for_evaluation` configuring the wrong flavor

### LINKED ISSUE #25073: [BUG] Post-import hooks in `configure_autologging_for_evaluation` configure the wrong flavor, silently disabling tracing during `mlflow.evaluate()`
> [!WARNING]
> Before submitting a PR, please make sure that:
> - A maintainer has triaged this issue and applied the `ready` label
> - This issue has no assignee
> - No duplicate PR exists
>
> PRs not meeting these requirements may be automatically closed.

### Issues Policy acknowledgement

- [x] I have read and agree to submit bug reports in accordance with the [issues policy](https://www.github.com/mlflow/mlflow/blob/master/ISSUE_POLICY.md)

### Where did you encounter this bug?

Local machine

### MLflow version

- Client: 3.15.2.dev0 (master, commit e6a3c7130936493adc148e111cfae0e9c0e712de)
- Tracking server: n/a (bug is client-side)

### System information

- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Windows 11 (10.0.26200)
- **Python version**: 3.10.18

### Describe the problem

`configure_autologging_for_evaluation` wraps every `mlflow.evaluate()`,
`mlflow.genai.evaluate()` and `optimize_prompts()` call. For each flavor it registers
a **post-import hook** that lazily enables tracing when the user's code imports that
flavor's library.

The hook closes over the loop variable `flavor`:

```python
for flavor in FLAVOR_TO_MODULE_NAME:
    original_config = global_config_snapshot.get(flavor, {}).copy()

    @autologging_conf_lock
    def _setup_autolog(module):
        autolog = get_autolog_function(flavor)     # <-- free variable
        ...                     original_config    # <-- free variable

    register_post_import_hook(_setup_autolog, FLAVOR_TO_MODULE_NAME[flavor], overwrite=True)
```

Python closures capture the *variable*, not its value, so all 17 hooks share a single
cell for `flavor`. `register_post_import_hook` fires a hook immediately **only** when
the module is already in `sys.modules`
([import_hooks/__init__.py#L104-L111](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/import_hooks/__init__.py#L104-L111));
otherwise it defers it. By the time a deferred hook runs, the loop has finished and
`flavor` holds its final value — `bedrock`.

Deferring is not an edge case, it is the stated design
([trace.py#L55-L58](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L55-L58)):

> *"Using post-import hook to configure the autologging lazily when the target flavor's
> module is imported, rather than configuring it immediately. This is because the
> evaluation code usually only uses a subset of the supported flavors, hence we want to
> avoid unnecessary overhead of configuring all flavors."*

**Expected behavior:** importing `langchain` during evaluation enables LangChain trace
autologging.

**Actual behavior:** importing `langchain` configures **bedrock** autologging.
LangChain tracing is never enabled. The same holds for `openai`, `dspy`, `llama_index`
and every other flavor except whichever one sorts last in `FLAVOR_TO_MODULE_NAME`.

The failure is silent twice over: no warning is emitted when the wrong flavor is
configured, and when the wrong flavor's library is not installed the resulting
`ImportError` is swallowed by the bare `except Exception` at
[trace.py#L85](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L85)
into a `_logger.debug`. The only user-visible symptom is *"my evaluation runs have no
traces."*

`ruff --select B023` independently flags lines 62, 65, 67 and 86.

### Secondary defect in the same function

The teardown at
[trace.py#L100-L107](https://github.com/mlflow/mlflow/blob/master/mlflow/models/evaluation/utils/trace.py#L100-L107)
tests the wrong dict:

```python
for module, hooks in new_import_hooks.items():
    if hooks is None:                        # never true
        _post_import_hooks.pop(module, None)
    else:
        _post_import_hooks[module] = original_import_hooks[module]   # may write None
```

`hooks` iterates the values of `new_import_hooks`, which are always the `_setup_autolog`
function, so the `is None` branch is dead and the else-branch writes `None` into
`_post_import_hooks` for every module that had no prior hook. The condition should test
`original_import_hooks[module]`.

Because `ImportHookFinder.find_spec` gates on key *presence*
([import_hooks/__init__.py#L308](https://github.com/mlflow/mlflow/blob/master/mlflow/utils/import_hooks/__init__.py#L308)),
a `None` value keeps MLflow intercepting imports of those modules for the remaining life
of the process — a redundant `find_spec` round-trip plus an `_ImportHookChainedLoader`
wrap on every subsequent import, after every `evaluate()` call. Section **C** of the
repro below demonstrates it.

Happy to fix both in one PR, or split them — whichever you prefer.

### Tracking information

```shell
N/A - not a tracking bug. Reproduces with any tracking URI.
```

### Code to reproduce issue

```python
"""Deferred post-import hooks configure the wrong flavor. No mocking, no ML libs needed."""

import inspect
import types

from mlflow.ml_package_versions import FLAVOR_TO_MODULE_NAME
from mlflow.models.evaluation.utils.trace import configure_autologging_for_evaluation
from mlflow.utils.autologging_utils import AUTOLOGGING_INTEGRATIONS
from mlflow.utils.import_hooks import _post_import_hooks

# Flavors whose library is typically NOT yet imported -> their hooks stay deferred,
# which is the path the lazy-configuration design exists to serve.
PROBES = ["langchain", "openai", "llama_index"]

print(f"Last flavor in FLAVOR_TO_MODULE_NAME: {list(FLAVOR_TO_MODULE_NAME)[-1]!r}\n")


def flavor_bound_to(hook):
    """Report which flavor a registered hook will actually configure."""
    fn = inspect.unwrap(hook)
    bound = inspect.signature(fn).parameters.get("flavor")
    if bound is not None and bound.default is not inspect.Parameter.empty:
        return bound.default
    return inspect.getclosurevars(fn).nonlocals.get("flavor")


print("A) Which flavor does each module's deferred hook actually configure?")
with configure_autologging_for_evaluation(enable_tracing=True):
    for flavor in PROBES:
        module = FLAVOR_TO_MODULE_NAME[flavor]
        for hook in _post_import_hooks.get(module) or []:
            actual = flavor_bound_to(hook)
            verdict = "OK" if actual == flavor else "WRONG"
            print(f"  [{verdict:5}] hook for module {module!r:20} -> configures {actual!r}")

print("\nB) Fire ONLY langchain's hook. Which flavor gets configured?")
before = set(AUTOLOGGING_INTEGRATIONS)
with configure_autologging_for_evaluation(enable_tracing=True):
    module = FLAVOR_TO_MODULE_NAME["langchain"]
    for hook in _post_import_hooks.get(module) or []:
        hook(types.ModuleType(module))
    touched = set(AUTOLOGGING_INTEGRATIONS) - before
    print(f"  importing {module!r} configured autologging for: {sorted(touched) or '(none)'}")
    print("  expected: ['langchain']")

print("\nC) Teardown residue in _post_import_hooks")
leaked = {
    FLAVOR_TO_MODULE_NAME[f]: _post_import_hooks.get(FLAVOR_TO_MODULE_NAME[f])
    for f in PROBES
    if FLAVOR_TO_MODULE_NAME[f] in _post_import_hooks
}
print(f"  keys left behind after the context manager exited: {leaked or '(none)'}")
print("  expected: (none)  -- ImportHookFinder gates on key presence")
```

### Stack trace

```
No exception is raised - this is a silent misconfiguration, which is what makes it
hard to notice in practice. The observed vs. expected output of the repro above is
included under "Other info / logs".

The one exception that does occur (when the wrongly-selected flavor's library is not
installed) is swallowed at mlflow/models/evaluation/utils/trace.py:85:

    except Exception:
        _logger.debug(f"Failed to update autologging config for {flavor}.", exc_info=True)

so it is invisible at the default log level.
```

### Other info / logs

```
$ python repro.py

Last flavor in FLAVOR_TO_MODULE_NAME: 'bedrock'

A) Which flavor does each module's deferred hook actually configure?
  [WRONG] hook for module 'langchain'            -> configures 'bedrock'
  [WRONG] hook for module 'openai'               -> configures 'bedrock'
  [WRONG] hook for module 'llama_index.core'     -> configures 'bedrock'

B) Fire ONLY langchain's hook. Which flavor gets configured?
  importing 'langchain' configured autologging for: ['bedrock', 'spark']
  expected: ['langchain']

C) Teardown residue in _post_import_hooks
  keys left behind after the context manager exited: {'langchain': None, 'openai': None, 'llama_index.core': None}
  expected: (none)  -- ImportHookFinder gates on key presence
```

Expected output (what you get once `flavor` / `original_config` are bound per-iteration
and the teardown checks the right dict):

```
A) Which flavor does each module's deferred hook actually configure?
  [OK   ] hook for module 'langchain'            -> configures 'langchain'
  [OK   ] hook for module 'openai'               -> configures 'openai'
  [OK   ] hook for module 'llama_index.core'     -> configures 'llama_index'

B) Fire ONLY langchain's hook. Which flavor gets configured?
  importing 'langchain' configured autologging for: ['langchain', 'spark']

C) Teardown residue in _post_import_hooks
  keys left behind after the context manager exited: (none)
```

Note: `'spark'` appears in both runs and is unrelated - `mlflow.langchain.autolog()`
also enables the Spark datasource integration. The signal is that `langchain` is
absent from the "actual" list and present in the "expected" one.

There are currently no tests referencing `configure_autologging_for_evaluation`.

### What component(s) does this bug affect?

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/scoring`: MLflow model serving, deployment tools, Spark UDFs
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [ ] `area/prompt`: MLflow prompt engineering features, prompt templates, and prompt management
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality
- [ ] `area/gateway`: MLflow AI Gateway client APIs, server, and third-party integrations
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/uiux`: Front-end, user experience, plotting
- [ ] `area/docs`: MLflow documentation pages

### PR DESCRIPTION
**Pull request by Koustav-github**
### Related Issues/PRs

Closes #25073

### What changes are proposed in this pull request?

Two related fixes in `configure_autologging_for_evaluation`, as separate commits per @PattaraS's steer on the issue.

**1. Bind `flavor` per-iteration (`mlflow/models/evaluation/utils/trace.py`)**

The post-import hook registered for each flavor closed over the loop variables `flavor` and `original_config` by reference:

```python
for flavor in FLAVOR_TO_MODULE_NAME:
    original_config = global_config_snapshot.get(flavor, {}).copy()

    @autologging_conf_lock
    def _setup_autolog(module):
        autolog = get_autolog_function(flavor)     # free variable
        ...                     original_config    # free variable

    register_post_import_hook(_setup_autolog, FLAVOR_TO_MODULE_NAME[flavor], overwrite=True)
```

`register_post_import_hook` fires a hook immediately only when the module is already in `sys.modules`; otherwise it defers it — the path this lazy design exists to serve. By the time a deferred hook ran, the loop had finished and both variables held their final values, so **every deferred hook configured the last flavor (`bedrock`) instead of its own**.

Importing `langchain` during evaluation therefore configured bedrock autologging, and LangChain tracing was never enabled. Silent in both directions: no warning, and a missing-library `ImportError` from the wrongly-selected flavor was swallowed into a `_logger.debug` at `trace.py:85`.

Fixed by binding both as default arguments. Also flagged by `ruff --select B023`.

**2. Restore import hooks from the correct dict**

The teardown iterated `new_import_hooks`, whose values are always the `_setup_autolog` function, so the `if hooks is None` branch was unreachable and the else-branch wrote `None` into `_post_import_hooks` for every module that had no prior hook. Since `ImportHookFinder.find_spec` gates on key *presence*, that residue kept MLflow intercepting imports of those modules for the rest of the process after every `evaluate()` call. Now iterates `original_import_hooks`, which holds the real prior state; that makes `new_import_hooks` write-only, so it is removed.

**Blast radius (issue ask #1)**

The behaviour change is narrower than it first looks: **the immediate-fire path was never broken.** When the module is already in `sys.modules`, the hook fires *inside* the loop iteration, where the loop variables still hold the right values. I verified this on both patched and unpatched `trace.py`. Only the deferred path changes.

| Flavors | Deferred behaviour before | After |
|---|---|---|
| `sklearn`, `pytorch`, `xgboost`, `lightgbm`, `statsmodels`, `spark`, `paddle`, `transformers` | configured `bedrock`; own flavor untouched | `autolog(disable=True)` — matches the documented intent |
| 16 traceable (`langchain`, `openai`, `dspy`, `llama_index`, `anthropic`, `gemini`, `crewai`, …) | configured `bedrock` | tracing correctly enabled ← the fix |
| `pytorch-lightning`, `keras`, `tensorflow`, `haystack`, `semantic_kernel` | `ImportError`, swallowed to debug | unchanged; only the logged flavor name differs |
| `pyspark.ml` | never registered (`is_autolog_supported` is False) | unchanged |

The only rows where the now-correct `autolog(disable=True)` is new are the non-traceable flavors in row 1, and only when the library is imported *during* evaluation. In the common case the library was imported before `evaluate()` was called (you needed it to build or load the model), the hook fired immediately, and `disable=True` already ran. I did not find a flavor where this looks like a regression risk.

### How is this PR tested?

- [x] New unit/integration tests

`tests/evaluate/test_autologging_config.py` — 13 tests; this function previously had none. **4 fail on `master`**:

| Test | On `master` |
|---|---|
| `test_deferred_hook_configures_its_own_flavor` | `['langchain', 'langchain']` != `['openai', 'langchain']` |
| `test_deferred_hook_uses_its_own_original_config` | `['from_langchain', 'from_langchain']` |
| `test_deferred_hook_disables_autolog_for_non_traceable_flavor` | enables tracing instead of disabling |
| `test_teardown_removes_hooks_that_had_no_prior_registration` | `{'_fake_module_openai': None, ...}` left behind |

Covering both paths requested on the issue:

- **enable path** — traceable flavor receives `log_traces=True, log_models=False, silent=True`
- **disable path** — non-traceable flavor receives `autolog(disable=True)`
- **`_should_enable_tracing` gate** — parametrized over real flavors with no mocking: `langchain`/`openai`/`dspy` → `True`, `sklearn`/`xgboost`/`transformers` → `False`
- **explicit opt-out** — flavor-level `log_traces: False` and global `mlflow.disable` still respected
- **teardown** — no `None` residue left behind, and pre-existing hooks restored intact

The tests keep hooks on the deferred path by pointing `FLAVOR_TO_MODULE_NAME` at module names absent from `sys.modules`, then firing the stored hooks directly. The two fake flavors are given different traceability, so the enable/disable tests also fail if the binding regresses. `_is_trace_autologging_supported` is deliberately left unpatched so the real `log_traces`-in-signature gate is exercised.

Each commit is independently green (11 tests at commit 1, 13 at commit 2), so the history stays bisectable.

**Regression check.** Ran `pytest tests/evaluate` on `master` and on this branch with identical flags: the set of failing tests is byte-identical (78 failed / 3 errors on both — pre-existing in my local Windows environment, mostly Spark/Java). No test changed status in either direction.

### Does this PR require documentation update?

- [x] No.

### Does this PR require updating the [MLflow Skills](https://github.com/mlflow/skills) repository?

- [x] No.

#### Is this a user-facing change?

- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Fixed a bug where auto-tracing was not enabled during `mlflow.evaluate()` and `mlflow.genai.evaluate()`. Post-import hooks captured the loop variable by reference, so every flavor's hook configured the last flavor instead of its own, and traces were silently missing for all flavors except one.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging
- [x] `area/evaluation`: MLflow model evaluation features, evaluation metrics, and evaluation workflows
- [x] `area/tracing`: MLflow Tracing features, tracing APIs, and LLM tracing functionality

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes

#### Is this PR a critical bugfix or security fix that should go into the next patch release?

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Minor releases are expected to contain larger changes, such as new features and improvements. Non-critical bug fixes and doc updates can be included as well. By default, your PR should target the next minor release.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Patch releases are typically only performed when there has been a major regression or bug in the latest release. For the sake of stability, your PR should not be included in a patch release unless it is a critical fix, or if the risk level of your PR is exceedingly low.

</details>

- [ ] This PR is critical and needs to be in the next patch release
- [x] This PR can wait for the next minor release

### CODE REVIEW

**Review by PattaraS — APPROVED** [review]
LGTM — verified locally. The new `tests/evaluate/test_autologging_config.py` passes 13/13, and reverting the per-iteration binding reproduces the bug exactly (`assert ['langchain', 'langchain'] == ['openai', 'langchain']`), so both new guards are genuine regression tests.

Both fixes are correct:
- **Late-binding closure:** binding `flavor`/`original_config` as default args snapshots each iteration's value (the hook fires as `hook(module)`, so the defaults apply). This is the real #25073 bug — deferred hooks were all configuring the last flavor (`bedrock`), silently disabling tracing for langchain/openai/dspy/etc.
- **Teardown:** the old `if hooks is None` branch was dead (values are always `_setup_autolog`), so it wrote `None` into `_post_import_hooks`; since `ImportHookFinder` gates on key presence, that left MLflow intercepting those imports for the process lifetime. Iterating `original_import_hooks` and popping on `None` is right, and the key sets are identical (both set in the same try block), so nothing is missed.

Nice test design too — deferred-path-only fakes, mixed-traceability flavor shapes that double as a wrong-flavor detector, and dedicated teardown guards. Thanks for the thorough fix.
